### ISSUE TITLE
Ellipsization is only enforced over line-wrapped lines

### ISSUE THREAD

**Issue by MyriaCore** [phuhl/linux_notification_center#98]
I've been noticing that the ellipsization code has a bug in it.

When working with multi-line bodies that are *also* line wrapped, I tend to get this weird behavior where the notification still stretches wayy lower than my chosen line limit. 

For example, the following notify-send command produces a gigantic notification:

```
notify-send 'Lorem Ipsum' "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent blandit molestie elit, et viverra sem commodo sed. Donec sed iaculis dolor. Proin vitae ex eu odio euismod lobortis. Etiam pretium consequat ligula vel commodo. In tempor magna ut metus posuere, nec elementum sapien suscipit. Fusce scelerisque condimentum metus, in scelerisque justo commodo sed. Fusce ac lorem euismod, porta magna ac, vehicula nulla.

Nullam posuere auctor varius. Curabitur in dui sit amet diam tempor porta ac eget libero. Sed quis felis sit amet ligula accumsan porttitor ut in enim. Quisque quam neque, vestibulum in volutpat varius, faucibus non tortor. Donec vitae neque sollicitudin, elementum ligula egestas, ultrices turpis. Donec porta vel sapien ac vestibulum. Donec rhoncus, tortor at volutpat consequat, orci enim interdum augue, et egestas leo lacus vel purus. Aliquam tempor vestibulum porta. Aliquam sed massa nec purus sollicitudin commodo. Donec vulputate massa quis eros pretium ullamcorper. Maecenas efficitur tortor at nisi lacinia, ac consequat ligula consequat. Aliquam consectetur metus ut tellus viverra, condimentum consequat mauris ornare. Nullam cursus orci id sodales efficitur. Pellentesque ultrices arcu congue sapien blandit, at sagittis risus dignissim.

Sed a dui vel ipsum pharetra aliquam. Phasellus sit amet fermentum urna. Sed porttitor mauris ante, eget tempus lacus viverra et. Sed et condimentum massa. Cras arcu lacus, venenatis ac nulla sed, mattis porta nisl. Aenean id massa sit amet augue elementum aliquam et vel enim. Donec mattis diam eu ipsum interdum, vitae mollis odio fringilla.

Ut non eros vitae sem accumsan tempus eu id dolor. Donec viverra pellentesque dui, at sagittis ligula condimentum vel. Donec eleifend dapibus nunc vel auctor. In ut quam eu massa commodo vestibulum. Morbi sit amet leo erat. Sed dignissim, mi eget consectetur pharetra, mi massa accumsan purus, sed commodo orci nisl vitae neque. Aliquam quis feugiat urna. Suspendisse neque enim, dapibus nec ullamcorper eget, consequat ornare sem.

Aliquam sit amet lorem mattis, fringilla mauris sit amet, aliquet ex. Mauris sit amet rutrum lectus. Nulla turpis neque, dictum sit amet sollicitudin vitae, mollis eget lacus. Donec at orci congue nibh dignissim molestie sit amet nec arcu. Maecenas elementum ex sed leo ornare pharetra. Integer laoreet finibus neque commodo efficitur. Nulla mollis at eros sit amet facilisis. Etiam id tempor orci, et hendrerit tellus. Quisque fringilla, risus sed commodo fermentum, dolor arcu lacinia turpis, eget aliquet mauris risus nec eros. Fusce at lacus sed odio convallis egestas ut a sem."
```

![image](https://user-images.githubusercontent.com/35076430/94607673-355cdf80-026a-11eb-847f-039d0c11c6c8.png)

As you can see, each *paragraph* is allowed to render, but is cut off at 5 lines. So, it appears that the 5 line ellipsization rule is only being enforced on a per-paragraph basis, rather than across the entire notification. 

It's important to note here that each **real line** in the text appears as a paragraph to us because of GTK's line wrapping, which breaks the text up. We need to make a distinction between real lines, which are the paragraph breaks, and line-wrapped lines, which are the lines that *we see* which make up the paragraph.

So, what's happening here, ultimately, is that ellipsization only appears to operate over *line-wrapped* lines, and not over the entire message. In this case, intended behavior would be for ellipsization operate once over the entire body, and to count both line-wrapped lines *and* real lines. This way, once 5 lines appear on the notification body, the ellipsization comes in to save your screen real-estate.

**Comment by MyriaCore** [comment]
It might be the case that ellipsization doesn't actually cover our use case, since GTK's ellipsization appears to be used to save horizontal screen real-estate rather than vertical, which is what we're using it for.

**Comment by MyriaCore** [comment]
It's been really difficult to find anything on this issue, but it appears to be that ellipsization really only happens over literal line breaks, not line-wrapping. I *did* find [this](https://mail.gnome.org/archives/gtk-app-devel-list/2007-June/msg00167.html) post in the gtk mailing list, however:

> > Does anyone know how to have text wrapping along with ellipsizing?
> > I want to have text wrapped to three lines and have ellipsis in the last
> > line if the text exceeds three lines.
> > 
> > I have used GtkLabel and set the properties for both wrapping and
> > ellipsizing.
> > The problem that I am facing is that if I try to wrap the text and
> > ellipsize, the text gets reduced to a single line and the ellipsis appear.
> 
> Yes, in ellipsize mode, every `paragraph' (run of text
> without any forced break) is ellipsized to fit the width.
> 
> You can probably achieve this ellipsization by using the
> more low-level functionality.  Iterating over
> PangoLayoutLines and rendering them yourself can be a bit
> too low-level, however, obtaining PangoLayoutLines and then
> inserting corresponding hard newlines into the original text
> and reflowing can give an acceptable result if lines are
> broken only between `words' (whatever it means for the
> script).  You can also try to measure the three lines, limit
> the rendering area accordingly, and then uses some graphic
> effect such as fade out instead of ellipsis.  Or something
> like that.

I've done some research about PangoLayoutLines to see if they could help us. Here are some notes:

- [GI.Gtk.Objects.Label labelGetLayout](https://www.stackage.org/haddock/lts-16.7/gi-gtk-3.0.33/GI-Gtk-Objects-Label.html#v:labelGetLayout) will return the [Layout](https://www.stackage.org/haddock/lts-16.7/gi-pango-1.0.22/GI-Pango-Objects-Layout.html) of a given label for us, which can help us see how many rendered lines there would be
- [GI.Pango.Objects.Layout layoutGetLineCount](https://www.stackage.org/haddock/lts-16.7/gi-pango-1.0.22/GI-Pango-Objects-Layout.html#v:layoutGetLineCount) returns the number of lines after line wrapping which we can use to check against our max line count (as configured by the user)
- If there are too many lines, [GI.Pango.Objects.Layout layoutGetLines](https://www.stackage.org/haddock/lts-16.7/gi-pango-1.0.22/GI-Pango-Objects-Layout.html#v:layoutGetLines) will return a list of [LayoutLines](https://www.stackage.org/haddock/lts-16.7/gi-pango-1.0.22/GI-Pango-Structs-LayoutLine.html), which won't be *directly* useful, but will allow us to get the LayoutLine of the last line we'll allow. 
- Now that we have the LayoutLine of the last line we want to allow, we need to figure out where the third-to-last character on that line is *in the text* so we can truncate there and add a `...`. We can do this with the following (I think):
  * <code><a href='https://www.stackage.org/haddock/lts-16.7/gi-pango-1.0.22/GI-Pango-Structs-LayoutLine.html#g:14'>the length</a> - 4</code> will get us the index of the third-to-last byte in the line
  * Add that to [the start index of the line](https://www.stackage.org/haddock/lts-16.7/gi-pango-1.0.22/GI-Pango-Structs-LayoutLine.html#g:17) to get the actual text index of this byte
  * Pray that 1 byte = 1 character because that's the only way this algorithm will work
- Finally, we change the character at the resulting index to a `.`, and then change the next 2 characters to a `.` as well, and truncate the text there. We'll call `labelSetText` or something to make the change. 

These notes do kinda assume we're fixing this by just inserting a `...` at the right location. We *could* try to do something fancier, but I'd be satisfied with this as a fix TBH.

**Comment by phuhl** [comment]
How about cutting the text at the first linebrake and insterting ...? That algorithm is /dirt simple/. Does produce mediocre results for very short first-lined messages, but better than totally broken...

**Comment by MyriaCore** [comment]
Yeah, fair enough. Although, if we need an emergency hotfix, then we should probably ship #110 early. ATM it *does* work perfectly for popups, but I still can't it to work nicely from inside the notification center. Regardless, in-center notis are less vulnerable than popups to this IMO, so we can ship a patch for the popups and then keep working on a patch for in-center notis.

**Comment by phuhl** [comment]
Ah, I might have fixed that issue you had, where you got only one line. Havn't pushed it yet, but it is a fix for #103 and #40 and makes sure, lines actually brake. It is not perfect as it now brakes lines by letters and not by words, but it's better than braking shit all over the place, I guess. So if you have time, you can check, if that really fixed PR #110 and if that is the case, we can merge it. Also, I am working on some features right now and plan to release a 2021 update (as you can see by the milestone I created). Would be nice, of course, if we could include the PR into that one.

Will push now (even though, not everything is done) so you can make use of the fix, I spoke of. It is commit 74cb7a19f7b30e17a35a25876545b8ff1c7d7d12

**Comment by MyriaCore** [comment]
I might not be able to get to this soon, but I hopefully will have some time later this week to try it out
